Web Inspector: Font Panel: CSS font property values marked !important don't get overridden when using the interactive editing controls.

francescorn opened the pull request:
#### e6f0e73090f834ff021ff7e702a179f11b827e08
<pre>
Web Inspector: Font Panel: CSS font property values marked !important don&apos;t get overridden when using the interactive editing controls.
<a href="https://bugs.webkit.org/show_bug.cgi?id=258975">https://bugs.webkit.org/show_bug.cgi?id=258975</a>

Review by NOBODY (OOPS!).

Fix the font-weight slider as the !important declaration causes the slider to to not slide and be overridden.

Address the case of having multiple of the same property of multiple font-weights and a !important declaration in the inline style.

FontStyles.js, writeFontVariation(): check if the target property exists, check if the existing target property contains !important,
check if the existing property is font-weight, update the value as !important if not already,
create the font-weight property if it doesn&apos;t exist, and set the CSSProperty raw value without marking it as !important.

Remove FixMe in CSSProperty.js, as there are intentional calls to the CSSProperty.value getter in use.

* Source/WebInspectorUI/UserInterface/Models/CSSProperty.js:
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles.prototype.writeFontVariation):
</pre>
----------------------------------------------------------------------
#### 3138cc78b5df4d3f4d3c67f9484639d93b0acb34
<pre>
Web Inspector: Font Panel: CSS font property values marked !important don&apos;t get overridden when using the interactive editing controls.
<a href="https://bugs.webkit.org/show_bug.cgi?id=258975">https://bugs.webkit.org/show_bug.cgi?id=258975</a>

Review by NOBODY (OOPS!).

Fix the font-weight slider as the !important declaration causes the slider to to not slide and be overridden.

Address the case of having multiple of the same property of multiple font-weights and a !important declaration in the inline style.

FontStyles.js, writeFontVariation(): check if the target property exists, check if the existing target property contains !important,
check if the existing property is font-weight, update the value as !important if not already,
create the font-weight property if it doesn&apos;t exist, and set the CSSProperty raw value without marking it as !important.

Remove FixMe in CSSProperty.js, as there are intentional calls to the CSSProperty.value getter in use.

* Source/WebInspectorUI/UserInterface/Models/CSSProperty.js:
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles.prototype.writeFontVariation):
</pre>
----------------------------------------------------------------------
#### f4a59075e299452a3cf03ea66fb35edc01acf215
<pre>
Web Inspector: Font Panel: CSS font property values marked !important don&apos;t get overridden when using the interactive editing controls.
<a href="https://bugs.webkit.org/show_bug.cgi?id=258975">https://bugs.webkit.org/show_bug.cgi?id=258975</a>

Review by NOBODY (OOPS!).

Fix the font-weight slider as the !important declaration causes the slider to to not slide and be overridden.

Address the case of having multiple of the same property of multiple font-weights and a !important declaration in the inline style.

FontStyles.js, writeFontVariation(): check if the target property exists, check if the existing target property contains !important,
check if the existing property is font-weight, update the value as !important if not already,
create the font-weight property if it doesn&apos;t exist, and set the CSSProperty raw value without marking it as !important.

Add FixMe for font-weight being renamed to font-variation-settings under style attributes.
Remove FixMe in CSSProperty.js, as there are intentional calls to the CSSProperty.value getter in use.

* Source/WebInspectorUI/UserInterface/Models/CSSProperty.js:
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles.prototype.writeFontVariation):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6f0e73090f834ff021ff7e702a179f11b827e08

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5218 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27850 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28815 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24330 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7033 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2624 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26864 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4046 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/22852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3565 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23840 "Found 2 new test failures: inspector/model/font-styles-write-variation.html, platform/mac/media/encrypted-media/fps-generateRequest.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29300 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24260 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24240 "Found 1 new test failure: inspector/model/font-styles-write-variation.html (failure)") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3640 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27769 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5079 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4114 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3972 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

francescorn committed:
Web Inspector: Font Panel: CSS font property values marked !important don't get overridden when using the interactive editing controls.
https://bugs.webkit.org/show_bug.cgi?id=258975

Review by NOBODY (OOPS!).

Fix the font-weight slider as the !important declaration causes the slider to to not slide and be overridden.

Address the case of having multiple of the same property of multiple font-weights and a !important declaration in the inline style.

FontStyles.js, writeFontVariation(): check if the target property exists, check if the existing target property contains !important,
check if the existing property is font-weight, update the value as !important if not already,
create the font-weight property if it doesn't exist, and set the CSSProperty raw value without marking it as !important.

Add FixMe for font-weight being renamed to font-variation-settings under style attributes.
Remove FixMe in CSSProperty.js, as there are intentional calls to the CSSProperty.value getter in use.

* Source/WebInspectorUI/UserInterface/Models/CSSProperty.js:
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles.prototype.writeFontVariation):

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -138,6 +138,13 @@ WI.FontStyles = class FontStyles
+        // FIXME: Only on the click of the slider for the first time you will notice that it does not move, but it will move
+        //  with any following clicks. From a developers perspective when sliding they may not notice it. That is caused
+        //  by creating the new newBlankProperty always creating overridden as false.
+        if (cssProperty.overridden || cssProperty.important)
```
This works, but only as a side-effect of the CSS property on the inline styles being overridden by an `!important` CSS property elsewhere in the CSS cascade.

It's always useful to build test cases to validate your changes. 

Here are two test cases where the fix is not working or creating unexpected output:

1. When there are existing CSS properties on the inline style that overwrite each other, only the first one gets used. If there is another one following it that's already `!important`, the changes made to this first CSS property are not reflected because they get overwritten by the latter one.

```html
<!DOCTYPE html>
<html>
<head>
    <title>font-size !important test case</title>
    <style type="text/css">
        body { font-family: system-ui; }
    </style>
</head>
<body style="font-weight: 100; font-weight: 500 !important;">
    Edit `Weight` in the Font sidebar panel of Web Inspector
</body>
</html>
```

Screen recording:

https://github.com/WebKit/WebKit/assets/105809798/8740f9df-53ff-4782-970f-bc5ba0b18ccf

2. When there are existing CSS properties on the inline style that overwrite each other, neither of them marked as `!important`, the first one gets changed and `!important` is appended to it. That's unexpected output. 

Web Inspector should write the value to the "winning" CSS property and not mark either one as `!important`.


```html
<!DOCTYPE html>
<html>
<head>
    <title>font-size !important test case</title>
    <style type="text/css">
        body { font-family: system-ui; }
    </style>
</head>
<body style="font-weight: 100; font-weight: 500;">
    Edit `Weight` in the Font sidebar panel of Web Inspector
</body>
</html>
```

Screen recording:

https://github.com/WebKit/WebKit/assets/105809798/5f54054e-d8de-41a5-a62c-2b7db8db8260

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -138,6 +138,13 @@ WI.FontStyles = class FontStyles
         const createIfMissing = true;
         let cssProperty = this._effectiveWritablePropertyForName(targetPropertyName, createIfMissing);
+
+        // FIXME: Only on the click of the slider for the first time you will notice that it does not move, but it will move
```
This code comment is not clear what it's about.

If it's a bug in a specific condition, it should be logged on Bugzilla instead with steps to reproduce, and, ideally, a small test case and a short screen recording if relevant.

rcaliman-apple reviewed (changes requested):
The fix works because it relies on a side-effect. That works against it in some scenarios. I left comments inline.

A more thorough fix requires tracking which CSS property that gets collected into `WI.FontStyles` is marked `!important` and remembering to apply `!important` to its corresponding CSS property when writing to the inline style.

(Aside: This whole idea of writing to the inline style is a workaround until we can figure out a solid way to write to an appropriate CSS rule in the cascade. That's a separate issue tracked in [Bug 250127](https://bugs.webkit.org/show_bug.cgi?id=250127))

Try to follow the flow of collecting CSS properties and their values in [FontStyles._calculateProperties()](https://github.com/WebKit/WebKit/blob/30f6d58c1ad8d7059ff4536c00081fb7520f3f1d/Source/WebInspectorUI/UserInterface/Models/FontStyles.js#L174-L204).  In particular, notice that [FontStyles._effectivePropertyValueForName()](https://github.com/WebKit/WebKit/blob/30f6d58c1ad8d7059ff4536c00081fb7520f3f1d/Source/WebInspectorUI/UserInterface/Models/FontStyles.js#L318-L321) only collects the value of the property, ignoring whether it was `important` or not.

When there are multiple candidate CSS properties to write to on the inline style, the correct one should be written to: the last declaration of that property or the pre-existing `!important` one if there is any.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/CSSProperty.js:
```diff
@@ -403,6 +402,8 @@ WI.CSSProperty = class CSSProperty extends WI.Object
 
     get priority() { return this._priority; }
 
+    set priority(priority) { this._priority = priority; }
```
`this._priority` is not used anywhere when updating the style text to be sent back to the backend. Setting this property has no effect on the value.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -138,7 +138,10 @@ WI.FontStyles = class FontStyles
+
         cssProperty.rawValue = targetPropertyValue;
+        if (this._nodeStyles._propertyNameToEffectivePropertyMap[cssProperty.name]?.important)
+            cssProperty.priority = "important";
```
Setting `WI.CSSProperty.priority = important` won't actually update the `rawValue` or send it to the backend (i.e. the inspected page) to reflect the style change. 

You were on the right track with concatenating `targetPropertyValue += " !important"`, but that needs to be done before assigning `WI.CSSProperty.rawValue`. 

That setter is the one that triggers the operations to update the style text and send it to the backend. 
Follow the operations called from here:

https://github.com/WebKit/WebKit/blob/54140a7b23efcb75955d08e7e7882371d1af3d99/Source/WebInspectorUI/UserInterface/Models/CSSProperty.js#L383-L397

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -138,7 +138,10 @@ WI.FontStyles = class FontStyles
         let cssProperty = this._effectiveWritablePropertyForName(targetPropertyName, createIfMissing);
+
         cssProperty.rawValue = targetPropertyValue;
+        if (this._nodeStyles._propertyNameToEffectivePropertyMap[cssProperty.name]?.important)
```
```diff
- this._nodeStyles._propertyNameToEffectivePropertyMap[cssProperty.name]?.important
+ this._nodeStyles._propertyNameToEffectivePropertyMap.get(cssProperty.name)?.important
```

francescorn replied:
Changed to:
`this._nodeStyles._propertyNameToEffectivePropertyMap.get(cssProperty.name)?.important`

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -330,7 +333,7 @@ WI.FontStyles = class FontStyles
         let properties = inlineCSSStyleDeclaration.visibleProperties;
 
-        cssProperty = properties.find(property => property.name === name);
+        cssProperty = properties.findLast(property => property.name === name && property.enabled);
```
`WI.CSSProperty.enabled` is not what you're looking for:

https://github.com/WebKit/WebKit/blob/54140a7b23efcb75955d08e7e7882371d1af3d99/Source/WebInspectorUI/UserInterface/Models/CSSProperty.js#L411-L412

It's likely `WI.CSSProperty.overridden`.

francescorn replied:
Changed to:
` cssProperty = properties.findLast(property => property.name === name && property.overridden);`

francescorn replied:
I tried && property.overridden, did you mean && !property.overridden?

rcaliman-apple reviewed (changes requested):
I don't know what source you used to record the latest screen recording, but it differs from the one submitted with this patch.

The code as submitted does not solve the original test case:

```html
<!DOCTYPE html>
<html>
<head>
    <title>font-size !important test case</title>
    <style type="text/css">
        body { font-family: system-ui; font-weight: 500 !important; }
    </style>
</head>
<body>
    Edit `Weight` in the Font sidebar panel of Web Inspector
</body>
</html>
```

Screen recording:

https://github.com/WebKit/WebKit/assets/105809798/4fa80b95-9451-45cc-bdc3-20ecceaa9a15

Additionally, here's another test case the current code doesn't handle well:

```html
<!DOCTYPE html>
<html>
<head>
    <title>font-size !important test case</title>
    <style type="text/css">
        body { font-family: system-ui; }
    </style>
</head>
<body style="font-weight: 100; font-weight: 500 !important; font-weight: 100">
    Edit `Weight` in the Font sidebar panel of Web Inspector
</body>
</html>
```

https://github.com/WebKit/WebKit/assets/105809798/0a7b58bb-99bb-422d-b6a8-188bdcf39545

francescorn committed:
Web Inspector: Font Panel: CSS font property values marked !important don't get overridden when using the interactive editing controls.
https://bugs.webkit.org/show_bug.cgi?id=258975

Review by NOBODY (OOPS!).

Fix the font-weight slider as the !important declaration causes the slider to to not slide and be overridden.

Address the case of having multiple of the same property of multiple font-weights and a !important declaration in the inline style.

FontStyles.js, writeFontVariation(): check if the target property exists, check if the existing target property contains !important,
check if the existing property is font-weight, update the value as !important if not already,
create the font-weight property if it doesn't exist, and set the CSSProperty raw value without marking it as !important.

Remove FixMe in CSSProperty.js, as there are intentional calls to the CSSProperty.value getter in use.

* Source/WebInspectorUI/UserInterface/Models/CSSProperty.js:
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles.prototype.writeFontVariation):

francescorn committed:
Web Inspector: Font Panel: CSS font property values marked !important don't get overridden when using the interactive editing controls.
https://bugs.webkit.org/show_bug.cgi?id=258975

Review by NOBODY (OOPS!).

Fix the font-weight slider as the !important declaration causes the slider to to not slide and be overridden.

Address the case of having multiple of the same property of multiple font-weights and a !important declaration in the inline style.

FontStyles.js, writeFontVariation(): check if the target property exists, check if the existing target property contains !important,
check if the existing property is font-weight, update the value as !important if not already,
create the font-weight property if it doesn't exist, and set the CSSProperty raw value without marking it as !important.

Remove FixMe in CSSProperty.js, as there are intentional calls to the CSSProperty.value getter in use.

* Source/WebInspectorUI/UserInterface/Models/CSSProperty.js:
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
(WI.FontStyles.prototype.writeFontVariation):
